clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java

```diff
@@ -52,13 +53,54 @@
 
     private final ConfigDef definition;
 
+    private static final String CONFIG_PROVIDERS_CONFIG = "config.providers";
+
+    /**
+     * Construct a configuration with a ConfigDef, the configuration properties, optional {ConfigProviders}.
+     *
+     * The originals is a name-value pair of config provider configs as well as configuration properties. The
+     * value of the configuration can be a variable as defined below or the actual value. This constructor will
+     * first instantiate the ConfigProviders using the config provider configs, then it will find all the
+     * variables in the values of the originals configurations, attempt to resolve the variables using the named
+     * ConfigProviders, and then parse and validate the configurations.
+     *
+     * ConfigProviders can be passed either as configs in the originals map or in the separate configProviderProps
+     * map. If config providers are passed in the configProviderProps any providers in originals map will be
+     * ignored. If ConfigProvider is not provided, the constructor will skip the variable substitution step and
+     * will simply validate and parse the supplied configuration.
+     *
+     * The config "config.providers" is reserved to list the ConfigProviders, "config.providers.<providerName>.class"
+     * is reserved to specify the class to be used to instantiate the providers and
+     * "config.providers.<providerName>.<variableName>" to specify any additional configs required by providers where
+     * <providerName> is the name of the new config provider and <variableName> are the variables required by the
+     * config provider.
+     *
+     * Variables have the form "${providerName:[path:]key}", where "providerName" is the name of a ConfigProvider,
+     * "path" is an optional string, and "key" is a required string. This variable is resolved by passing the "key"
+     * and optional "path" to a ConfigProvider with the specified name, and the result from the ConfigProvider is
+     * then used in place of the variable. Variables that cannot be resolved by the AbstractConfig constructor will
+     * be left unchanged in the configuration.
+     *
+     *
+     * @param definition the definition of the configurations; may not be null
+     * @param originals the name-value pairs of the configuration; may not be null
+     * @param configProviderProps the map of properties of config providers which will be instantiated by the constructor to
+     *        resolve any variables in {@code originals}; may be null or empty
```

Comment:
We should specify whether the keys in this map have the `config.providers.` prefix, like they would in the `originals`.